### PR TITLE
iterative ref: generate_xy_plane

### LINKED ISSUE #14: iterative ref: generate_xy_plane
see how meshgrid and generate coordinates functions used in https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/simulate.py#L96

### PR DESCRIPTION
**Pull request by jedyeo**

### CODE REVIEW

**Comment by geoffwoollard** [comment]
@jedyeo 

No, don't try to install coords. Re-write it in reconstructSPI from https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/coords.py#L7

It's something I wrote that used to be in a module of compSPI, but was stashed away where there was a cleanup. Keep it in here and it will be refactored into compSPI down the road.

**Comment by thisFreya** [comment]
@jedyeo @geoffwoollard FYI while working on #15 I realized that map_coordinates is effectively the only place we use xy_plane. So, its shape should reflect the parameter formats of that function. To this end, I've changed its shape requirement to (3, n_pix**2), i.e.,
```
[ x1 x2 x3 x4 x5 ]
[ y1 y2 y3 y4 y5 ]
[ z1 z2 z3 z4 z5 ]
```
And adjusted tests, etc. in this branch accordingly. 

Also of importance is the ordering of xy_plane. Say we have a 2x2x2 pixel grid. So, xy_plane contains 4 vectors. Then, if I were to plot the vectors of xy_plane I would get something like this:
image
I.e., vectors pointing to four equidistant points on the grid. Now, what we should enforce (and why) is what order these vectors are stored in. The output of map_coordinates is the values at the rotated coordinates, so in this case it would be 4 scalars **f** = [f1, f2, f3, f4] that would need to fit in to each of the points these vectors point to. The simplest (computationally) way to get the desired output would be to reshape **f**, i.e., slice = f.reshape(2,2). *This only works if the vectors used to generate **f** point to the "points" in the 2x2 slice array that reshape puts the values in.* 

How you could enforce this (in a test) is to generate a 4-entry xy_plane as I've been discussing and directly interpreting the x,y coordinates of each vector as entries in an array, putting increasing values between 0 and 4 in each slot, and reshaping:
```
[ -.5  .5 -.5 .5 ]    [ 0 1 ]    
[ -.5 -.5  .5 .5 ] -> [ 2 3 ] -> [ 0 1 2 3 ]
[   0   0   0  0 ]    
```
As long as this is *consistent* and either maps in order to an (x,y) grid or a (y,x) grid we should be able to avoid any errors.

**Comment by thisFreya** [comment]
@jedyeo Another point here: Planes should be **centered** about (x,y,z) = (0,0,0). This is to facilitate simple, unambiguous rotation characteristics (again, since rotation is the purpose of these planes they should be catered to it). I've updated the other post accordingly.

**Comment by thisFreya** [comment]
@jedyeo After having functionality clarified in generate_slices I was able to add some hardcoded tests so I did that.

**Comment by geoffwoollard** [comment]
> @jedyeo @geoffwoollard FYI while working on #15 I realized that map_coordinates is effectively the only place we use xy_plane. So, its shape should reflect the parameter formats of that function.

Also the xy coordinates are used when the coordinates get rotated by the 3x3 rotation matrix

**Comment by thisFreya** [comment]
> > FYI while working on #15 I realized that map_coordinates is effectively the only place we use xy_plane. So, its shape should reflect the parameter formats of that function.
> 
> Also the xy coordinates are used when the coordinates get rotated by the 3x3 rotation matrix

@geoffwoollard Very true, I should have spoken more carefully. I standby the shape change as the coordinate rotation after the fact can admit either shape, while the map_coordinates function is strict. However, either form is just a transpose away.

**Comment by geoffwoollard** [comment]
> Also of importance is the ordering of xy_plane. Say we have a 2x2x2 pixel grid. So, xy_plane contains 4 vectors. Then, if I were to plot the vectors of xy_plane I would get something like this:

If it's a 2x2 plane then the coordinates of one dim would be [-1,0]. So in the (3,n**2) we would have
```
[[-1, -1, 0, 0],
[-1, 0, -1, 0,]]
```

As I recall, in the notebooks there is some swapping around of rows sometimes (like so)
```
[,[-1, 0, -1, 0,],
[-1, -1, 0, 0]]
```

So that the results of flattening an array match. e.g. the values of `arr_3d.flatten()`, which will be `n**3`, match the coordinates of `coords_3d.flatten()`

This is useful in case there can be speed ups with binary masks where not all the values / coordinates are operated on. I definitely did this in 2D... but and will likely do this in 3D once there is a custom linear interpolation function that is not `map_coordinates`.

you want to be able to do things like`arr_3d[binary_mask_3d] = some_values`, and have this correspond to the right voxels in `coords_3d[:,binary_mask_3d]`.

So please don't go too far afield from when is in the notebooks I've written

**Comment by geoffwoollard** [comment]
> How you could enforce this (in a test) is to generate a 4-entry xy_plane as I've been discussing and directly interpreting the x,y coordinates of each vector as entries in an array, putting increasing values between 0 and 4 in each slot, and reshaping:
> 
> ```
> [ -.5  .5 -.5 .5 ]    [ 0 1 ]    
> [ -.5 -.5  .5 .5 ] -> [ 2 3 ] -> [ 0 1 2 3 ]
> [   0   0   0  0 ]    
> ```
> 
> As long as this is _consistent_ and either maps in order to an (x,y) grid or a (y,x) grid we should be able to avoid any errors.

I don't quite understand what you're getting at here. But I think you are on the right track. Perhaps you should just come up with something that passes tests and then I can double check it works out.

**Comment by geoffwoollard** [comment]
> @jedyeo After having functionality clarified in generate_slices I was able to add some hardcoded tests so I did that.

Look closely at how the coordinates (coords_3d of shape (3,n**3)) are changed by adding back in n//2 https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/simulate.py#L111 so that they agree with the convention of `map_coordinates`.  This makes the "top corner" of a arr_3d, be the origin `[0,0,0]`, and not `[-n//2,-n//2,-n//2]`. And the actual "center" (first pixel of right half) pixel in `arr_3d` correspond to coordinate `[n//2,n//2,n//2]`.

**Comment by geoffwoollard** [comment]
> @jedyeo
> 
> No, don't try to install coords. Re-write it in reconstructSPI from https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/coords.py#L7
> 
> It's something I wrote that used to be in a module of compSPI, but was stashed away where there was a cleanup. Keep it in here and it will be refactored into compSPI down the road.

Reminder that you'll want to do this sooner rather than later. And write some tests for `coords.coords_n_by_d`

It's a very simple function. Just 

```
def coords_n_by_d(coords_1d=None, N=None, d=3):
    if coords_1d is not None and N is None:
        pass
    elif coords_1d is None and N is not None:
        assert N % 2 == 0, "N must be even"
        coords_1d = np.arange(-N // 2, N // 2)
    else:
        assert False, "use either N or coords_1d, not both"

    if d == 2:
        X = np.meshgrid(coords_1d, coords_1d)
    elif d == 3:
        X = np.meshgrid(coords_1d, coords_1d, coords_1d)
    coords = np.zeros((X[0].size, d))
    for di in range(d):
        coords[:, di] = X[di].flatten()
    if d == 3:
        coords[:, [0, 1]] = coords[:, [1, 0]]

    return coords
```

**Comment by jedyeo** [comment]
@geoffwoollard I've adapted a simplified version of your code from `coords`. This generates a 'centered' plane. e.g the discrete values on an axis range from `-n_pix / 2` to `n_pix / 2 - 1`

File: reconstructSPI/iterative_refinement/expectation_maximization.py [hunk 1]
```diff
@@ -334,12 +336,16 @@ def generate_xy_plane(n_pix):
...
         """
-        # See how meshgrid and generate coordinates functions used
-        # https://github.com/geoffwoollard/compSPI/blob/stash_simulate/src/simulate.py#L96
+        axis_pts = np.arange(-n_pix // 2, n_pix // 2)
+        grid = np.meshgrid(axis_pts, axis_pts)
+
+        xy_plane = np.zeros((3, n_pix**2))
+
+        for d in range(2):
+            xy_plane[d, :] = grid[d].flatten()
```
Comment (geoffwoollard) [inline reconstructSPI/iterative_refinement/expectation_maximization.py:347]: We can change the initial shape to be that which is returned. Then we don't need the transpose at the end.

**Review by geoffwoollard — APPROVED** [review]

**Comment by geoffwoollard** [comment]
Fixes https://github.com/compSPI/reconstructSPI/issues/14

**Pull request merged**